### PR TITLE
Add composite action to CI

### PR DESCRIPTION
**Pull request by eike-hass**
# Description of change
Split up our `build-and-test` workflow into reusable actions using github composite actions https://docs.github.com/en/actions/creating-actions/creating-a-composite-action

## Type of change
Add an `x` to the boxes that are relevant to your changes.

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [X] Enhancement (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

## How the change has been tested
ran Jobs in github CI

## Change checklist
Add an `x` to the boxes that are relevant to your changes.

- [X] I have followed the contribution guidelines for this project
- [X] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -307,5 +190,6 @@ jobs:
         run: wasm-pack test --node
         working-directory: bindings/wasm
 
-      - name: Stop sccache server
-        run: sccache --stop-server || true
+      - uses: './.github/actions/rust/sccache/stop-sccache-ubuntu-macos'
```
Comment (eike-hass) [inline .github/workflows/build-and-test.yml:193]: unfortunately we can't use `post` steps with composite actions

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -43,99 +41,34 @@ jobs:
...
-          scoop install sccache
-          echo "${HOME}/scoop/apps/sccache/current" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
-
-      - name: Start sccache
-        run: |
-          sccache --start-server
-          sccache -s
 
+      - uses: './.github/actions/rust/sccache/install-and-start-sccache-macos'
+        name: install-and-start-sccache-macos
```
Comment (cycraig) [inline .github/workflows/build-and-test.yml:69]: Improve name?

```suggestion
      - name: Setup sccache (MacOS)
        uses: './.github/actions/rust/sccache/install-and-start-sccache-macos'
```

File: .github/workflows/build-and-test.yml [hunk 3]
```diff
@@ -149,8 +82,13 @@ jobs:
           command: test
           args: --all --release
 
-      - name: Stop sccache server
-        run: sccache --stop-server || true
+      - uses: './.github/actions/rust/sccache/stop-sccache-windows'
+        name: stop-sccache-windows
```
Comment (cycraig) [inline .github/workflows/build-and-test.yml:86]: Improve name?
```suggestion
      - name: Stop sccache (Windows)
        uses: './.github/actions/rust/sccache/stop-sccache-windows'
```

File: .github/workflows/build-and-test.yml [hunk 4]
```diff
@@ -149,8 +82,13 @@ jobs:
...
+        if: matrix.os == 'windows-latest'
+
+      - uses: './.github/actions/rust/sccache/stop-sccache-ubuntu-macos'
+        name: stop-sccache-ubuntu-macos
```
Comment (cycraig) [inline .github/workflows/build-and-test.yml:90]: Improve name?

```suggestion
      - name: Stop sccache (Ubuntu/MacOS)
        uses: './.github/actions/rust/sccache/stop-sccache-ubuntu-macos'
```

File: .github/workflows/build-and-test.yml [hunk 5]
```diff
@@ -199,83 +137,28 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
-      - name: Install stable toolchain
-        uses: actions-rs/toolchain@v1
+      - uses: './.github/actions/utils/get-current-time-ubuntu-macos'
+        name: get-current-time-ubuntu-macos
```
Comment (cycraig) [inline .github/workflows/build-and-test.yml:141]: Improve name?

```suggestion
      - name: Get current date (Ubuntu/MacOS)
        uses: './.github/actions/utils/get-current-time-ubuntu-macos'
```

File: .github/workflows/build-and-test.yml [hunk 6]
```diff
@@ -199,83 +137,28 @@ jobs:
...
-          path: ${{ matrix.sccache-path }}
-          key: ${{ runner.os }}-sccache-${{ github.job }}-${{ hashFiles('**/Cargo.lock') }}
-          restore-keys: |
-            ${{ runner.os }}-sccache-${{ github.job }}-
-            ${{ runner.os }}-sccache-
-
-      - name: Install sccache (ubuntu-latest)
+        
+      - uses: './.github/actions/rust/sccache/install-and-start-sccache-ubuntu'
+        name: install-and-start-sccache-ubuntu
```
Comment (cycraig) [inline .github/workflows/build-and-test.yml:160]: Improve name?

```suggestion
      - name: Setup sccache (Ubuntu)
        uses: './.github/actions/rust/sccache/install-and-start-sccache-ubuntu'
```

File: .github/workflows/build-and-test.yml [hunk 7]
```diff
@@ -307,5 +190,6 @@ jobs:
         run: wasm-pack test --node
         working-directory: bindings/wasm
 
-      - name: Stop sccache server
-        run: sccache --stop-server || true
+      - uses: './.github/actions/rust/sccache/stop-sccache-ubuntu-macos'
+        name: stop-sccache-ubuntu-macos
```
Comment (cycraig) [inline .github/workflows/build-and-test.yml:194]: Improve name?

```suggestion
      - name: Stop sccache (Ubuntu/MacOS) 
        uses: './.github/actions/rust/sccache/stop-sccache-ubuntu-macos'
```

**Review by cycraig — APPROVED** [review]
Looks good to me. Only nits on naming things from my side really.

The lack of conditionals or post-steps in composite actions is disappointing, as we now have multiple composite actions for each platform. This is still worthwhile for re-using steps across jobs and workflows, however.

File: .github/actions/rust/rust-setup/action.yml [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  job:
+    description: ''
+    required: true
+  current-date:
+    description: ''
+    required: true
+  target-cache-path:
+    description: ''
+    required: false
+    default: target
```
[the remark this answers is not in the document]
Comment (eike-hass) [reply .github/actions/rust/rust-setup/action.yml:19]: We can't remove them unfortunately. Will add short descriptions.

File: .github/actions/utils/get-current-date-windows/action.yml [hunk 9]
```diff
@@ -0,0 +1,8 @@
+name: 'get-current-time-windows'
```
Comment (cycraig) [inline .github/actions/utils/get-current-date-windows/action.yml:1]:
```suggestion
name: 'get-current-date-windows'
```

**Pull request merged**